node/src/Pos/Block/Logic/Header.hs

```diff
@@ -224,52 +223,52 @@ classifyHeaders inRecovery headers = do
-    :: forall ssc m.
-       (DB.MonadBlockDB ssc m, WithLogger m, MonadError Text m, HasConfiguration, HasNodeConfiguration)
-    => NonEmpty HeaderHash  -- ^ Checkpoints; not guaranteed to be
-                            --   in any particular order
+getHeadersFromManyTo ::
+       forall ssc m.
+       ( DB.MonadBlockDB ssc m
+       , WithLogger m
+       , MonadError Text m
+       , HasConfiguration
+       , HasNodeConfiguration
+       )
+    => NonEmpty HeaderHash -- ^ Checkpoints; not guaranteed to be
+                           --   in any particular order
     -> Maybe HeaderHash
     -> m (NewestFirst NE (BlockHeader ssc))
 getHeadersFromManyTo checkpoints startM = do
     logDebug $
         sformat ("getHeadersFromManyTo: "%listJson%", start: "%build)
                 checkpoints startM
-    validCheckpoints <- noteM "Failed to retrieve checkpoints" $
-        nonEmpty . catMaybes <$>
-        mapM (DB.blkGetHeader @ssc) (toList checkpoints)
-    tip <- GS.getTip
-    unless (all ((/= tip) . headerHash) validCheckpoints) $
-        throwError "Found checkpoint that is equal to our tip"
-    let startFrom = fromMaybe tip startM
-        isCheckpoint bh =
-            any (\c -> bh ^. headerHashG == c ^. headerHashG) validCheckpoints
-        parentIsCheckpoint bh =
-            any (\c -> bh ^. prevBlockL == c ^. headerHashG) validCheckpoints
-        whileCond bh = not (isCheckpoint bh)
-    headers <- noteM "Failed to load headers by depth" . fmap (_Wrapped nonEmpty) $
-        DB.loadHeadersByDepthWhile whileCond recoveryHeadersMessage startFrom
-    let newestH = headers ^. _Wrapped . _neHead
-        oldestH = headers ^. _Wrapped . _neLast
-    logDebug $
-        sformat ("getHeadersFromManyTo: retrieved headers, oldest is "
-                % build % ", newest is " % build) oldestH newestH
-    if parentIsCheckpoint oldestH
-    then pure headers
-    else do
-        logDebug $ "getHeadersFromManyTo: giving headers in recovery mode"
-        inMainCheckpoints <-
-            noteM "Filtered set of valid checkpoints is empty" $
-            nonEmpty <$> filterM GS.isBlockInMainChain (toList validCheckpoints)
-        logDebug $ "getHeadersFromManyTo: got checkpoints in main chain"
-        let lowestCheckpoint =
-                maximumBy (comparing getEpochOrSlot) inMainCheckpoints
-            loadUpCond _ h = h < recoveryHeadersMessage
-        up <- GS.loadHeadersUpWhile lowestCheckpoint loadUpCond
-        res <- note "loadHeadersUpWhile returned empty list" $
-            _Wrapped nonEmpty (toNewestFirst $ over _Wrapped (drop 1) up)
-        logDebug $ "getHeadersFromManyTo: loaded non-empty list of headers, returning"
-        pure res
+
+    tip <- DB.getTipHeader @ssc
+    let tipHash = headerHash tip
+
+    -- This filters out invalid/unknown checkpoints also.
+    inMainCheckpoints <-
+        noteM "no checkpoints are in the main chain" $
+        nonEmpty <$> filterM GS.isBlockInMainChain (toList checkpoints)
```

Comment:
`isBlockInMainChain` is faster than `blkGetHeader` I take it? So we get a speedup here?

Reply:
We perform one traversal versus two (loadHeadersByDepthWhile and then loadHeadersUpWhile).

Reply:
Sure but in the normal (non recovery) case there's only one traversal and it includes no block deserialization. Now in every case there is block deserialization.

Reply:
1. You don't seem to understand the usecase and how this function was designed previously, general question is: it's still faster or almost the same speed, i believe.
2. It doesn't really matter because I'll fix `loadHeadersUpWhile` in a moment.

Reply:
> You don't seem to understand the usecase and how this function was designed previously

Sure, I'd believe this.

> it's still faster or almost the same speed, i believe.

But this is harder to believe.

> It doesn't really matter because I'll fix loadHeadersUpWhile in a moment

Ok, that should make it better.